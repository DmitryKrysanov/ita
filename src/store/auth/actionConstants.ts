export const SIGNIN_SUCCESSFUL = "SIGNIN_SUCCESSFUL";
export const SIGNIN_ERROR = "SIGNIN_ERROR";
export const SIGNOUT = "SIGNOUT";
