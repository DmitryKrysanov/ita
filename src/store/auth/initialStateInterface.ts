export interface InitialStateInterface {
    userReducer: UserInterface
}

export interface UserInterface {
    uid: string | null,
    token: string | null,
    error: null
}