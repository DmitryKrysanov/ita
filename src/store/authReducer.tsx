import * as actionConstants from '../store/actions-constants';

const initialState = {
  token: null,
  id: null,
  error: null
}

const authReducer = (state = initialState, action: any) => {
  console.log('auth reducer');
  switch (action.type) {
    case actionConstants.AUTH:
      return {
        ...state,
        error: null
      }
    case actionConstants.AUTH_SUCCESSFUL:
      return {
        ...state,
        token: action.payload.token,
        id: action.payload.id
      }
    case actionConstants.AUTH_ERROR:
        return {
          ...state,
          error: action.payload
        }
    default: return state;
  }

}

export default authReducer;