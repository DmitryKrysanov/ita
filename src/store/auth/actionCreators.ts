import { SIGNIN_ERROR, SIGNIN_SUCCESSFUL, SIGNOUT } from "./actionConstants";
import { Firebase } from "../../components/App/Auth/firebase.config";
import Cookies from "js-cookie";

export const successSignin = (user: any) => {
  return {
    type: SIGNIN_SUCCESSFUL,
    payload: user
  };
};

const errorSignin = (err: Error) => {
  return {
    type: SIGNIN_ERROR,
    payload: err
  };
};

export const signInUser = (email: string, password: string) => {
  return async (dispatch: any) => {
    try {
      const response = await Firebase.auth().signInWithEmailAndPassword(
        email,
        password
      );

      if (response.user) {
        console.dir(response.user.getIdTokenResult());
        const token = await response.user
          .getIdTokenResult(true)
          .then(idToken => idToken.token);
        const refreshToken = response.user.refreshToken;
        Cookies.set("token", token, { expires: 1 / 24});
        Cookies.set("refreshToken", refreshToken);
        Cookies.set("uid", response.user.uid);
        dispatch(successSignin({ ...response.user, token }));
      }
    } catch (err) {
      dispatch(errorSignin(err.message));
    }
  };
};
export const signOutUser = () => {
  return (dispatch: any) => {
    Firebase.auth()
      .signOut()
      .then(() => {
        Cookies.remove("token");
        Cookies.remove("refreshToken");
        Cookies.remove("uid");
        dispatch({
          type: SIGNOUT
        });
      });
  };
};
