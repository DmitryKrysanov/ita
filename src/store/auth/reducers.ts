import { combineReducers } from 'redux'
import { SIGNIN_ERROR, SIGNIN_SUCCESSFUL, SIGNOUT} from './actionConstants'
import {appReducer} from "../reduces";
import { UserInterface } from "./initialStateInterface";
import Cookies from "js-cookie";


let token = Cookies.get('token') || Cookies.get('refreshToken');
const id = Cookies.get('uid');

export const initialState: UserInterface = {
  uid: id || null,
  token: token || null,
  error: null
};

export const userReducer = (state = initialState, action: any) => {
    console.log(action.type);
    switch (action.type) {
        case SIGNIN_SUCCESSFUL:
            return {
                ...state,
                token: action.payload.token || state.token,
                uid: action.payload.uid || state.uid,
                error: null
            };
        case SIGNIN_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case SIGNOUT:
            return {
                ...state = initialState,
                token: null,
                uid: null
            };
        default:
            return state
    }
};

const rootReducer = combineReducers({
    userReducer,
    appReducer
});

export default rootReducer;