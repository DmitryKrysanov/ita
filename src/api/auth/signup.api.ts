import { SignupInterface, SignupFullDataInterface } from '@components/App/Auth/Signup/SignupInterface';
import { auth, db } from '../../components/App/Auth/firebase.config';
import { signInUser } from '../../store/auth/actionCreators';


export const authSignup = (userData: SignupInterface) => {
  return auth
    .createUserWithEmailAndPassword(userData.email, userData.password)
    .then((data: any) => {console.dir(data);signInUser(userData.email, userData.password); return data.user})
    .catch(() => null);
};

export const authSignupFullData = (fullUserData: SignupFullDataInterface) => {
  const time = new Date().getTime();
  const threeMonth = 7889400000;
  return db
    .collection('users')
    .doc(fullUserData.id)
    .set({
      email: fullUserData.email,
      name: fullUserData.name,
      group: fullUserData.group,
      time: {
        startDate: new Date(time),
        startTime: time,
        endDate: new Date(time + threeMonth),
        endTime: time + threeMonth
      },
      tasks: {
        resolvedTasks: 0,
        failedTasks: 0
      }
    })
    .then(() => true)
    .catch(() =>  false);
};