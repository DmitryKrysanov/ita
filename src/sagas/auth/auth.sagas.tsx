import { takeLatest, call, fork, put } from 'redux-saga/effects';
import * as actionTypes from '../../store/auth/actionConstants';
import * as actions from '../../store/auth/actionCreators';
import * as API from '../../api/auth/signup.api';
import { AuthDataInterface, SignupFullDataInterface } from '@components/App/Auth/Signup/SignupInterface';
import { createUserInformation } from '../../components/App/Auth/AuthUtils';

function* getAuth(authData: any) {
  const userToken: string = yield authData.getIdToken()
    .then((token: string) => token );
  const userId: string = authData.uid;
  const refreshToken: string = authData.refreshToken;
  const auth: AuthDataInterface = {
    token: userToken,
    refreshToken: refreshToken,
    uid: userId
  }
  return auth;
}

function* signupWithFullUserData(data: SignupFullDataInterface) {
  const isSuccess: boolean = yield call(API.authSignupFullData, {
    uid: data.uid,
    email: data.email,
    name: data.name,
    group: data.group,
  });
  return isSuccess;
}

function* saveDataInState(userAuthData: any) {
  const storageUserData: AuthDataInterface = {
    token: userAuthData.token,
    uid: userAuthData.uid,
    refreshToken: userAuthData.refreshToken
  }
  yield put(actions.saveUserAuthData(
    storageUserData
  ));
  
  if (!storageUserData.token || !storageUserData.uid || !storageUserData.refreshToken) { return; }
  createUserInformation(storageUserData.token, storageUserData.refreshToken, storageUserData.uid);
}

function* signupUser(action: any) {
  try{
    const user = yield call(API.authSignup, {
      email: action.payload.email,
      password: action.payload.password,
      name: action.payload.name,
      group: action.payload.group,
    });

    if (!user) { return; }

    const userAuthData: AuthDataInterface = yield call(getAuth, user);
    const isSuccess = yield call(signupWithFullUserData, { uid: userAuthData.uid, ...action.payload });

    if(!isSuccess) { return; }

    yield call(saveDataInState, {...userAuthData, error: null, isAuth: true });
  }catch(e){
    // error handle in future
  }
}

function* watchSignupUser() {
  yield takeLatest(actionTypes.AUTH_SIGNUP, signupUser);
}

const authSagas = [
  fork(watchSignupUser)
];

export default authSagas;