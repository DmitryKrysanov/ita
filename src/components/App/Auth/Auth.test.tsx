import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import AuthPopUp from "./Auth";
import configureStore, { MockStore } from "redux-mock-store";
import { initialState } from "../../../store/auth/authReducer";
import { Provider } from "react-redux";
import "@testing-library/jest-dom/extend-expect";

const mockStore = configureStore([]);
afterEach(cleanup);
describe("AuthPopUp", () => {
  let store: MockStore;
  let component: any;
  beforeEach(() => {
    store = mockStore({
      authReducer: { ...initialState }
    });
    store.dispatch = jest.fn();
    component = render(
      <Provider store={store}>
        <AuthPopUp />
      </Provider>
    );
  });

  it("should render with Auth component", () => {
    expect(component).toMatchSnapshot();
  });

  it("should close the window by clicking on icon", () => {
    const { baseElement } = component;
    const closeIcon = baseElement.querySelector("i");
    expect(closeIcon).toBeInTheDocument();
  });

  it("should sign out by clicking on button", () => {
    const { baseElement } = component;
    const SignoutBtn = baseElement.querySelector(".signOut");
    expect(SignoutBtn).toBeInTheDocument();
    fireEvent.click(SignoutBtn);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
