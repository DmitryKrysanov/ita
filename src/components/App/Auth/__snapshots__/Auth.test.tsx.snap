// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthPopUp should render with Auth component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 0px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    >
      <div />
    </div>
    <div
      class="MuiDialog-root"
      role="presentation"
      style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper"
        role="none presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
          role="dialog"
        >
          <div
            class="makeStyles-root-1"
          >
            <header
              class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorDefault MuiPaper-elevation4"
            >
              <div
                aria-label="scrollable auto tabs example"
                class="MuiTabs-root"
              >
                <div
                  class="MuiTabs-scrollable"
                  style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
                />
                <div
                  class="MuiTabs-scroller MuiTabs-scrollable"
                  style="margin-bottom: 0px;"
                >
                  <div
                    class="MuiTabs-flexContainer"
                    role="tablist"
                  >
                    <button
                      aria-controls="scrollable-auto-tabpanel-0"
                      aria-selected="true"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
                      id="scrollable-auto-tab-0"
                      role="tab"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiTab-wrapper"
                      >
                        SIGN IN
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                    <button
                      aria-controls="scrollable-auto-tabpanel-1"
                      aria-selected="false"
                      class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
                      id="scrollable-auto-tab-1"
                      role="tab"
                      tabindex="0"
                      type="button"
                    >
                      <span
                        class="MuiTab-wrapper"
                      >
                        SIGN UP
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <span
                    class="PrivateTabIndicator-root-80 PrivateTabIndicator-colorPrimary-81 MuiTabs-indicator"
                    style="left: 0px; width: 0px;"
                  />
                </div>
              </div>
            </header>
            <div
              aria-labelledby="scrollable-auto-tab-0"
              class="MuiTypography-root MuiTypography-body1"
              id="scrollable-auto-tabpanel-0"
              role="tabpanel"
            >
              <div
                class="MuiBox-root MuiBox-root-115"
              >
                <form
                  class="sc-bxivhb kUFzYx"
                >
                  <div
                    class="sc-bdVaJa dBgsZc"
                  >
                    <div>
                      <div
                        class="MuiFormControl-root MuiTextField-root"
                      >
                        <label
                          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                          data-shrink="false"
                          for="EmailSignin"
                          id="EmailSignin-label"
                        >
                          Email
                        </label>
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiOutlinedInput-input"
                            id="EmailSignin"
                            name="email"
                            type="email"
                            value=""
                          />
                          <fieldset
                            aria-hidden="true"
                            class="PrivateNotchedOutline-root-176 MuiOutlinedInput-notchedOutline"
                            style="padding-left: 8px;"
                          >
                            <legend
                              class="PrivateNotchedOutline-legend-177"
                              style="width: 0.01px;"
                            >
                              <span>
                                ​
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="sc-bdVaJa dBgsZc"
                  >
                    <div>
                      <div
                        class="MuiFormControl-root MuiTextField-root"
                      >
                        <label
                          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                          data-shrink="false"
                          for="PasswordSignin"
                          id="PasswordSignin-label"
                        >
                          Password
                        </label>
                        <div
                          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiOutlinedInput-input"
                            id="PasswordSignin"
                            name="password"
                            type="password"
                            value=""
                          />
                          <fieldset
                            aria-hidden="true"
                            class="PrivateNotchedOutline-root-176 MuiOutlinedInput-notchedOutline"
                            style="padding-left: 8px;"
                          >
                            <legend
                              class="PrivateNotchedOutline-legend-177"
                              style="width: 0.01px;"
                            >
                              <span>
                                ​
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </div>
                  </div>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained submit MuiButton-containedPrimary"
                    tabindex="0"
                    type="submit"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      SignIn
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </form>
              </div>
            </div>
            <div
              aria-labelledby="scrollable-auto-tab-1"
              class="MuiTypography-root MuiTypography-body1"
              hidden=""
              id="scrollable-auto-tabpanel-1"
              role="tabpanel"
            />
          </div>
          <i
            class="sc-ifAKCX kKXnou fas fa-times"
          />
          <button
            class="signOut"
          >
            Sign out
          </button>
        </div>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
